[visionOS] Introduce experimental controls API for spatial images

yoelhawa opened the pull request:
#### 77f02397ad6c9e34e93941a9967b2d6c21e4c1a7
<pre>
[visionOS] Introduce experimental controls API for spatial images
<a href="https://bugs.webkit.org/show_bug.cgi?id=281347">https://bugs.webkit.org/show_bug.cgi?id=281347</a>
&lt;<a href="https://rdar.apple.com/134260482">rdar://134260482</a>&gt;

Reviewed by Ryosuke Niwa.

Introduce an experimental feature flag for a declarative API for
viewing spatial photos on the web, by adding a `controls` attribute on
the `&lt;img&gt;` element. When enabled and used, this should add a button
only to a spatial photo that when tapped brings the image into
fullscreen to be viewed spatially.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
Add the experimental feature flag.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/SpatialImageControls.cpp: Added.
(WebCore::SpatialImageControls::spatialImageControlsElementIdentifier):
(WebCore::SpatialImageControls::spatialImageControlsButtonIdentifier):
(WebCore::SpatialImageControls::hasSpatialImageControls):
(WebCore::SpatialImageControls::spatialImageControlsHost):
(WebCore::SpatialImageControls::isSpatialImageControlsButtonElement):
(WebCore::SpatialImageControls::shouldHaveSpatialControls):
(WebCore::SpatialImageControls::addShadowRootForSpatialImageControls):
(WebCore::SpatialImageControls::ensureSpatialControls):
(WebCore::SpatialImageControls::handleEvent):
(WebCore::SpatialImageControls::destroySpatialImageControls):
(WebCore::SpatialImageControls::updateSpatialImageControls):
* Source/WebCore/dom/SpatialImageControls.h: Added.
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::attributeChanged):
(WebCore::HTMLImageElement::didAttachRenderers):
(WebCore::HTMLImageElement::didAddUserAgentShadowRoot):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/shadow/spatialImageControls.css: Added.
(#if defined(ENABLE_SPATIAL_IMAGE_CONTROLS) &amp;&amp; ENABLE_SPATIAL_IMAGE_CONTROLS):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::notifyFinished):
* Source/WebCore/rendering/RenderImage.cpp:
Implement functionality for `controls` attribute on `&lt;img&gt;` element.

* LayoutTests/TestExpectations:
* LayoutTests/fast/images/resources/spatial.heic: Added.
* LayoutTests/fast/images/spatial-image-controls-expected.html: Added.
* LayoutTests/fast/images/spatial-image-controls.html: Added.
Add new layout tests.

Canonical link: <a href="https://commits.webkit.org/285453@main">https://commits.webkit.org/285453@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77f3a6d43f3a4374d986fbba2196913e79b8e2a1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/72708 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/52133 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/25506 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/23942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/74823 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59938 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/23742 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/76898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/23942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/75775 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/47149 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/62585 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/76898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/43799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/20051 "Passed tests") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/22271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/65841 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/65652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/20410 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/78570 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/71964 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16954 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/19539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/65634 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/17002 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/62593 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/64911 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/13210 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/6857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/93743 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11166 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47931 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/2718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20623 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/48998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/50293 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/48743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1065,3 +1065,7 @@
 #define ENABLE_UNFAIR_LOCK 1
 #endif
+
+#if PLATFORM(VISION) && __VISION_OS_VERSION_MAX_ALLOWED >= 20400
```
Not sure this gets you the EWS builders.
(apparently not)

yoelhawa replied:
Noted. Setting the new layout tests to skip for now then until EWS catches up

etiennesegonzac commented on LayoutTests/fast/images/spatial-image-controls.html:
```diff
@@ -0,0 +1,20 @@
+    setTimeout(()=>{
+        if (window.testRunner)
+            testRunner.notifyDone();
+    }, 200)
```
I'm sure the hardcoded timeout came from another test, but could we avoid it by
- waiting for both images to be loaded
- then waiting on `UIHelper.renderingUpdate()` (1 frame to paint + 1 tick making sure your queued task runs)
?

yoelhawa replied:
Oh perfect, didn't know that `UIHelper.renderingUpdate` existed!

etiennesegonzac commented on Source/WebCore/dom/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,219 @@
+        Ref controlLayer = HTMLDivElement::create(document.get());
+        controlLayer->setIdAttribute(spatialImageControlsElementIdentifier());
+        controlLayer->setAttributeWithoutSynchronization(HTMLNames::contenteditableAttr, falseAtom());
+        controlLayer->setAttributeWithoutSynchronization(HTMLNames::styleAttr, "position: relative;"_s);
```
Just curious: could this be in the UA stylesheet you're adding?
(related: do we need the extra container right now?)

yoelhawa replied:
The extra container (with `position: relative`) is necessary for correct positioning of the button.

To include it in the UA stylesheet I add I'd have to move the style block out of being within the `div` so it could apply to the `div`, and I ran into issues when removing the controls (when the `controls` attribute gets removed) when the shadow root had two immediate children (`style` and `div`). I could keep looking for a workaround if this is bad form though?

etiennesegonzac replied:
nah, that makes sense!

rniwa commented on LayoutTests/fast/images/spatial-image-controls.html:
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html><!-- webkit-test-runner [ SpatialImageControlsEnabled=true ] -->
```
Can we turn this into a ref test?

rniwa commented on LayoutTests/TestExpectations:
```diff
@@ -48,6 +48,7 @@ fast/forms/textarea/ios [ Skip ]
 fast/forms/watchos [ Skip ]
 fast/images/text-recognition [ Skip ]
 fast/images/mac [ Skip ]
+fast/images/spatial-image-controls.html [ Skip ]
```
Wouldn't this mean the test is always skipped? We need to enable it on visionOS, right?

yoelhawa replied:
In an earlier comment it was pointed out that EWS won't compile this until future SDK so it'll always fail on EWS, so it's been suggested offline that the test skip for now and later PR to un-skip on visionOS once EWS updates. Unless there's a better way?

rniwa replied:
I see. That seems fine then.

rniwa commented on Source/WebCore/dom/SpatialImageControls.h:
```diff
@@ -0,0 +1,51 @@
+class HTMLImageElement;
+class HTMLElement;
+
+namespace SpatialImageControls {
```
I don't think this file belongs to Source/WebCore/dom. It belongs in Source/WebCore/html/shadow directory.

rniwa commented on LayoutTests/fast/images/spatial-image-controls-expected.html:
```diff
@@ -0,0 +1,32 @@
+        width: 200px;
+        position: relative;
+    }
+    #controls {
```
This effectively copies UA stylesheet's rule... that's not great but I don't think there is any other way.

rniwa commented on LayoutTests/fast/images/spatial-image-controls-expected.html:
```diff
@@ -0,0 +1,32 @@
+<!DOCTYPE html>
```
We need to test how this feature interacts with ImageOverlay.
See tests in LayoutTests/fast/images/text-recognition.
We need to add a test scenario in which both this feature and text recognition feature kicks in.

yoelhawa replied:
The tests in LayoutTests/fast/images/text-recognition all fail on visionOS when I tried running them (except the ones about text recognition NOT running). I assume text recognition isn't supported on visionOS, or at least not in the simulator where all visionOS layout tests run.

rniwa commented on Source/WebCore/html/HTMLImageElement.cpp:
```diff
@@ -438,6 +442,10 @@ void HTMLImageElement::attributeChanged(const QualifiedName& name, const AtomStr
 #endif
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+    SpatialImageControls::updateSpatialImageControls(*this);
```
This is not an appropriate place to do this. This function runs every time an attribute is mutated on an image element.

rniwa replied:
Please add it under `AttributeNames::controlsAttr` in the above switch.

rniwa commented on Source/WebCore/html/HTMLImageElement.cpp:
```diff
@@ -501,6 +509,15 @@ void HTMLImageElement::didAttachRenderers()
         renderImage->setImageSizeForAltText();
 }
 
+void HTMLImageElement::didAddUserAgentShadowRoot(ShadowRoot& shadowRoot)
```
I don't think we want to override this function.
Instead, ensureSpatialControls should call `ensureUserAgentShadowRoot()` and add the spatial control.

yoelhawa replied:
Removed the override of didAddUserAgentShadowRoot

rniwa commented on Source/WebCore/dom/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,219 @@
+    if (!element.document().settings().spatialImageControlsEnabled())
+        return false;
+
+    bool hasSpatialcontrolsAttribute = element.hasAttributeWithoutSynchronization(HTMLNames::controlsAttr);
```
Is this spec'ed at all?

yoelhawa replied:
Not at this time, which is why it's behind an experimental feature flag. That would of course have to change before this ever would become on by default.

rniwa commented on Source/WebCore/dom/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,219 @@
+
+    auto* cachedImage = element.cachedImage();
+    if (!cachedImage)
+        return false;
```
This isn't right. cached image doesn't get set immediately when src attribute, etc... changes. We need to detect when an image gets loaded via ImageLoader. Maybe we wanna add a code to do this in `ImageLoader::didUpdateCachedImage`?

rniwa replied:
How does that work? You're calling `updateSpatialImageControls` in just `HTMLImageElement::attributeChanged`.

yoelhawa replied:
Sorry misspoke earlier and didn't realize you already replied before I deleted my comment.

Refactored things a bit so `updateSpatialImageControls` gets called in exactly two places: `HTMLImageElement::attributeChanged` and `ImageLoader::notifyFinished` (the cached image will always be fully loaded by the latter). The former is just to catch if the `controls` attribute is added or removed by javascript after the image has already loaded. In the event that the image hasn't yet loaded by the time a JS call decides to add the `controls` attribute, `updateSpatialImageControls` should gracefully fail when called by `attributeChanged` but it will be called again by `notifyFinished`.

rniwa commented on LayoutTests/fast/images/spatial-image-controls.html:
```diff
@@ -0,0 +1,38 @@
+</style>
+<body>
+    <div>
+        <img controls src="resources/spatial.heic" id="spatial">
```
We need to test what happens when the image is loaded lazily (loading=lazy):
https://html.spec.whatwg.org/multipage/urls-and-fetching.html#lazy-loading-attribute

yoelhawa replied:
Note: Haven't added test yet but quick manual test shows it works (with network inspector to confirm image was in fact lazily loading)

rniwa commented on Source/WebCore/dom/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,219 @@
+    if (!host)
+        return nullptr;
+
+    return hasSpatialImageControls(*host) ? host : nullptr;
```
We definitely want to add a test where spatial image appears within an author defined shadow tree.

yoelhawa replied:
Added.

rniwa commented on Source/WebCore/dom/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,219 @@
+    if (shouldHaveSpatialControls(imageElement) && !hasSpatialImageControls(imageElement)) {
+        imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+            RefPtr element = weakElement.get();
+            element->ensureUserAgentShadowRoot();
```
We should invoke SpatialImageControls::ensureSpatialControls here instead of overriding didAddUserAgentShadowRoot

yoelhawa replied:
Removed the override of didAddUserAgentShadowRoot

rniwa commented on Source/WebCore/html/HTMLImageElement.idl:
```diff
@@ -34,6 +34,9 @@
     [CEReactions=NotNeeded, Reflect] attribute DOMString useMap;
     [CEReactions=NotNeeded, Reflect] attribute boolean isMap;
+#if defined(ENABLE_SPATIAL_IMAGE_CONTROLS) && ENABLE_SPATIAL_IMAGE_CONTROLS
+    [CEReactions=NotNeeded, Reflect] attribute boolean controls;
```
Can we also add EnabledBySetting=SpatialImageControlsEnabled. Otherwise, this attribute will be exposed on HTMLImageElement whenever we build with this option.

rniwa commented on Source/WebCore/html/shadow/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,214 @@
+
+    imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+        RefPtr element = weakElement.get();
+        RefPtr shadowRoot = element->shadowRoot();
```
We need to check the nullity of element. Element could be dead between the time this callback is scheduled and when it's called so add:
```cpp
if (!element)
    return;
```

rniwa commented on Source/WebCore/html/shadow/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,214 @@
+    imageElement.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { imageElement }] {
+        RefPtr element = weakElement.get();
+        RefPtr shadowRoot = element->shadowRoot();
+        ASSERT(shadowRoot);
```
We should probably sanity check that this shadow root's mode is UserAgent.

rniwa commented on Source/WebCore/html/shadow/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,214 @@
+void destroySpatialImageControls(HTMLElement& element)
+{
+    element.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { element }] {
+        RefPtr protectedElement = weakElement.get();
```
Ditto about checking nullity of `protectedElement` here.

rniwa commented on Source/WebCore/html/shadow/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,214 @@
+            RefPtr htmlElement = dynamicDowncast<HTMLElement>(node.releaseNonNull());
+            if (!htmlElement)
+                return;
+            RELEASE_ASSERT_WITH_SECURITY_IMPLICATION(htmlElement->getIdAttribute() == spatialImageControlsElementIdentifier());
```
I don't think you can release-assert this. UA shadow tree could be hosting image overlay for example.

rniwa replied:
I suggest we look for an element with the matching ID instead like this:
```cpp
if (RefPtr element = shadowRoot->getElementById(spatialImageControlsElementIdentifier()))
    element->remove();
```

rniwa commented on Source/WebCore/html/shadow/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,214 @@
+        return;
+
+    element.protectedDocument()->checkedEventLoop()->queueTask(TaskSource::InternalAsyncTask, [weakElement = WeakPtr { element }] {
+        RefPtr element = weakElement.get();
```
Need to check the nullity of element here as well.

rniwa commented on Source/WebCore/html/HTMLImageElement.cpp:
```diff
@@ -430,6 +434,11 @@ void HTMLImageElement::attributeChanged(const QualifiedName& name, const AtomStr
     }
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+    case AttributeNames::controlsAttr:
+        SpatialImageControls::updateSpatialImageControls(*this);
```
We need to do this when the image gets updated as well but that could happen in a follow-up.

yoelhawa replied:
Isn't it sufficient that `updateSpatialImageControls` also gets called by `ImageLoader::notifyFinished` for image updates?

rniwa replied:
Ah, I missed that. Yeah, that'll work.

rniwa commented on Source/WebCore/html/shadow/SpatialImageControls.cpp:
```diff
@@ -0,0 +1,208 @@
+    bool shouldHaveControls = shouldHaveSpatialControls(element);
+    bool hasControls = hasSpatialImageControls(element);
+
+    if (shouldHaveControls && !hasControls) {
```
No curly braces around a single line statement.

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -470,6 +474,12 @@ void ImageLoader::notifyFinished(CachedResource& resource, const NetworkLoadMetr
+
+#if ENABLE(SPATIAL_IMAGE_CONTROLS)
+    RefPtr imageElement = dynamicDowncast<HTMLImageElement>(element());
+    if (imageElement)
```
Declare inside if:
```
if (RefPtr imageElement = ...
```

Commit:
[visionOS] Introduce experimental controls API for spatial images
https://bugs.webkit.org/show_bug.cgi?id=281347
<rdar://134260482>

Reviewed by Ryosuke Niwa.

Introduce an experimental feature flag for a declarative API for
viewing spatial photos on the web, by adding a `controls` attribute on
the `<img>` element. When enabled and used, this should add a button
only to a spatial photo that when tapped brings the image into
fullscreen to be viewed spatially.

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformEnableCocoa.h:
Add the experimental feature flag.

* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/dom/SpatialImageControls.cpp: Added.
(WebCore::SpatialImageControls::spatialImageControlsElementIdentifier):
(WebCore::SpatialImageControls::spatialImageControlsButtonIdentifier):
(WebCore::SpatialImageControls::hasSpatialImageControls):
(WebCore::SpatialImageControls::spatialImageControlsHost):
(WebCore::SpatialImageControls::isSpatialImageControlsButtonElement):
(WebCore::SpatialImageControls::shouldHaveSpatialControls):
(WebCore::SpatialImageControls::addShadowRootForSpatialImageControls):
(WebCore::SpatialImageControls::ensureSpatialControls):
(WebCore::SpatialImageControls::handleEvent):
(WebCore::SpatialImageControls::destroySpatialImageControls):
(WebCore::SpatialImageControls::updateSpatialImageControls):
* Source/WebCore/dom/SpatialImageControls.h: Added.
* Source/WebCore/html/HTMLButtonElement.cpp:
(WebCore::HTMLButtonElement::defaultEventHandler):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::attributeChanged):
(WebCore::HTMLImageElement::didAttachRenderers):
(WebCore::HTMLImageElement::didAddUserAgentShadowRoot):
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLImageElement.idl:
* Source/WebCore/html/shadow/spatialImageControls.css: Added.
(#if defined(ENABLE_SPATIAL_IMAGE_CONTROLS) && ENABLE_SPATIAL_IMAGE_CONTROLS):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::ImageLoader::notifyFinished):
* Source/WebCore/rendering/RenderImage.cpp:
Implement functionality for `controls` attribute on `<img>` element.

* LayoutTests/TestExpectations:
* LayoutTests/fast/images/resources/spatial.heic: Added.
* LayoutTests/fast/images/spatial-image-controls-expected.html: Added.
* LayoutTests/fast/images/spatial-image-controls.html: Added.
Add new layout tests.

Canonical link: https://commits.webkit.org/285453@main
